tracker/src/main/kotlin/org/dhis2/tracker/relationships/data/RelationshipsRepository.kt

```diff
@@ -0,0 +1,219 @@
+        } ?: false
+    }
+
+    protected fun getTeiAttributesForRelationship(
+        teiUid: String?,
+        relationshipConstraint: RelationshipConstraint?,
+        relationshipCreationDate: Date?,
+    ): List<Pair<String, String>> {
+        //Get list of ordered attributes
+        val trackedEntityAttributesUids = when {
+
+            //When there are  attributes defined in the constraint
+            relationshipConstraint?.trackerDataView()?.attributes()?.isNotEmpty() == true -> {
+                relationshipConstraint.trackerDataView()?.attributes()
+            }
+
+            //If there is a program defined in the constraint
+            relationshipConstraint?.program() != null -> {
+                val programUid = relationshipConstraint.program()?.uid()
+                d2.programModule().programTrackedEntityAttributes()
+                    .byProgram().eq(programUid)
+                    .byDisplayInList().isTrue
+                    .blockingGet().mapNotNull {
+                        it.trackedEntityAttribute()?.uid()
+                    }
+            }
+
+            //If there is no program then we get the trackedEntity type attributes
+            relationshipConstraint?.trackedEntityType()?.uid() != null -> {
+                val teiTypeUid = relationshipConstraint.trackedEntityType()?.uid()
+                d2.trackedEntityModule().trackedEntityTypeAttributes()
+                    .byTrackedEntityTypeUid().eq(teiTypeUid)
+                    .byDisplayInList().isTrue.blockingGet().mapNotNull {
+                        it.trackedEntityAttribute()?.uid()
+                    }
+            }
+
+            else -> {
+                emptyList()
+            }
+
+        }
+
+        //Get a list of Pair<DisplayName, value>
+        val attributes = trackedEntityAttributesUids?.mapNotNull { attributeUid ->
+            val fieldName = d2.trackedEntityModule().trackedEntityAttributes()
+                .uid(attributeUid).blockingGet()
+                ?.displayFormName()
+
+            val value = d2.trackedEntityModule().trackedEntityAttributeValues()
+                .value(attributeUid, teiUid!!).blockingGet()?.userFriendlyValue(d2)
+            if (fieldName != null && value != null) {
+                Pair(fieldName, value)
+            } else {
+                null
+            }
+        } ?: emptyList()
+
+        return attributes.ifEmpty {
+            val teiTypeUid = relationshipConstraint?.trackedEntityType()?.uid()
+            val teiTypeName = d2.trackedEntityModule().trackedEntityTypes()
+                .uid(teiTypeUid).blockingGet()?.name() ?: ""
+            listOf(Pair(teiTypeName, relationshipCreationDate?.toUi() ?: ""))
+        }
+    }
+
+    protected fun getEventValuesForRelationship(
+        eventUid: String?,
+        relationshipConstraint: RelationshipConstraint?,
+        relationshipCreationDate: Date?,
+    ): List<Pair<String, String>> {
```

Comment:
There is a very handy way of making this return something meaningful. Just from reading it the user does not know what it that Pair of strings.

`typeAlias StageDisplayName = String`
`typeAlias RelationshipUiCreationDate = String`
`protected fun getEventValuesForRelationship(...): List<Pair<StageDisplayName , RelationshipUiCreationDate >>`